packages/ckeditor5-list/src/todolist/todolistediting.ts

```diff
@@ -392,6 +398,47 @@ export default class TodoListEditing extends Plugin {
 	}
 }
 
+/**
+ * Returns an upcast converter for to-do list items.
+ */
+function todoListItemUpcastConverter(): GetCallback<UpcastElementEvent> {
+	return ( evt, data, conversionApi ) => {
+		const { writer, schema } = conversionApi;
+
+		if ( !data.modelRange ) {
+			return;
+		}
+
+		// Group to-do list items by their listItemId attribute to ensure that all items of the same list item have the same checked state.
+		const groupedItems = Array
+			.from( data.modelRange.getItems( { shallow: true } ) )
+			.filter( ( item ): item is Element =>
+				item.getAttribute( 'listType' ) === 'todo' && schema.checkAttribute( item, 'listItemId' )
+			)
+			.reduce( ( acc, item ) => {
+				const listItemId = item.getAttribute( 'listItemId' ) as string;
+				const items = acc.get( listItemId ) || [];
+
+				items.push( item );
+				acc.set( listItemId, items );
+
+				return acc;
+			}, new Map<string, Array<Element>>() );
```

Comment:
Can we use the util helper from here: https://ckeditor.com/docs/ckeditor5/latest/api/module_list_list_utils_model.html#function-getAllListItemBlocks ?

Reply:
@niegowski Added in this commit https://github.com/ckeditor/ckeditor5/pull/16785/commits/ae81fc1f120e8c152d106c7629ab356302f83c6f